topology/intel/dmic-nhlt.c

```diff
@@ -0,0 +1,1632 @@
+	 * topology if it is non-zero, otherwise use default length.
+	 */
+	/* unmute_ramp related stuff is not used in nhlt at all */
+	if (dmic_prm[di].unmute_ramp_time)
+		unmute_ramp_time_ms = dmic_prm[di].unmute_ramp_time;
+	else
+		unmute_ramp_time_ms = LOGRAMP_TIME_MS;
+
+	if (unmute_ramp_time_ms < LOGRAMP_TIME_MIN_MS ||
+	    unmute_ramp_time_ms > LOGRAMP_TIME_MAX_MS) {
+		tplg_pp_debug("dmic_set_config(): Illegal ramp time = %d",
+			      unmute_ramp_time_ms);
+		ret = -1;
+		goto out;
+	}
+
+	if (di >= DMIC_HW_FIFOS) {
+		tplg_pp_debug("dmic_set_config(): dai->index exceeds number of FIFOs");
+		ret = -1;
+		goto out;
+	}
+
+	if (dmic_prm[di].num_pdm_active > DMIC_HW_CONTROLLERS) {
+		tplg_pp_debug("dmic_set_config(): the requested PDM controllers count exceeds platform capability");
+		ret = -1;
+		goto out;
+	}
+
+	switch (dmic_prm[di].fifo_bits) {
+	case 0:
+	case 16:
+	case 32:
+		break;
+	default:
+		tplg_pp_debug("dmic_set_config(): fifo_bits EINVAL");
+		ret = -1;
+		goto out;
+	}
+
+	/* Match and select optimal decimators configuration for FIFOs A and B
+	 * paths. This setup phase is still abstract. Successful completion
+	 * points struct cfg to FIR coefficients and contains the scale value
+	 * to use for FIR coefficient RAM write as well as the CIC and FIR
+	 * shift values.
+	 */
+	find_modes(&modes_a, dmic_prm[0].fifo_fs, di);
+	if (modes_a.num_of_modes == 0 && dmic_prm[0].fifo_fs > 0) {
+		tplg_pp_debug("dmic_set_config(): No modes found found for FIFO A");
+		ret = -1;
+		goto out;
+	}
+
+	find_modes(&modes_b, dmic_prm[1].fifo_fs, di);
+	if (modes_b.num_of_modes == 0 && dmic_prm[1].fifo_fs > 0) {
+		tplg_pp_debug("dmic_set_config(): No modes found for FIFO B");
+		ret = -1;
+		goto out;
+	}
+
+	match_modes(&modes_ab, &modes_a, &modes_b);
+	ret = select_mode(&cfg, &modes_ab);
+	if (ret < 0) {
+		tplg_pp_debug("dmic_set_config(): select_mode() failed");
+		ret = -1;
+		goto out;
+	}
+
+	/* Struct reg contains a mirror of actual HW registers. Determine
+	 * register bits configuration from decimator configuration and the
+	 * requested parameters.
+	 */
+	ret = configure_registers(&cfg);
+	if (ret < 0) {
+		tplg_pp_debug("dmic_set_config(): cannot configure registers");
+		ret = -1;
+		goto out;
+	}
+
+	/* put some default values here */
+	dmic_vendor_blob.gateway_attributes = 1;
+	dmic_vendor_blob.ts_group[0] = 4294967057;
+	dmic_vendor_blob.ts_group[1] = 4294967040;
+	dmic_vendor_blob.ts_group[2] = 4294967295;
+	dmic_vendor_blob.ts_group[3] = 4294967295;
+	dmic_vendor_blob.ts_group[3] = 4294967295;
+	dmic_vendor_blob.clock_on_delay = 16;
+
+#ifdef DEBUG_DMIC_VENDOR_BLOB
+	print_dmic_vendor_blob();
+	print_dmic_vendor_blob_as_bytes();
+#endif /* DEBUG_DMIC_VENDOR_BLOB*/
+
+out:
+	return ret;
+}
+
+static int set_dmic_data(struct tplg_pre_processor *tplg_pp,
+			 snd_config_t *dai_cfg)
+{
+	snd_config_iterator_t i, next;
+	snd_config_t *n;
+	const char *id;
+	long int_val = 0;
+	int driver_version = 1;
+	int io_clk = 19200000;
```

Comment:
is this generic? This wouldn't work with 'older' platforms with a 24MHz clk.

Reply:
removed default values.